socketshark/receiver.py

```diff
@@ -29,70 +29,71 @@ def __init__(self, shark, redis_receiver):
 
         # We use a special channel to pass the stop message to the reader.
         self._stop_channel = self.redis_receiver.channel('_internal')
+        self._stop = False  # Stop flag
 
     def _channel(self, name):
         return self.redis_channel_prefix + name
 
     async def ping_handler(self):
         ping_interval = self.redis_settings['ping_interval']
-        if not ping_interval:
+        ping_timeout = self.redis_settings['ping_timeout']
+        if not ping_interval or not ping_timeout:
             return
+
         latency = 0
+        ping = wait = None
+
         try:
             while True:
+                # Sleep before pings
+                await asyncio.sleep(ping_interval - latency)
+
                 self.shark.log.debug('redis ping')
+
                 start_time = time.time()
-                # Until https://github.com/aio-libs/aioredis/issues/249 is
-                # fixed, we simply post a message on the ping channel.
-                ping_channel = self.redis_channel_prefix + '_socketshark_ping'
-                ping = self.redis.publish(ping_channel, '')
-                timeout_handler = asyncio.ensure_future(
-                        self.ping_timeout_handler(ping))
-                try:
-                    await ping
-                except asyncio.CancelledError:  # Cancelled by timeout handler
-                    break
 
-                latency = time.time() - start_time
-                self.shark.log.debug('redis pong', latency=round(latency, 3))
+                ping = self.redis.ping()
+                wait = asyncio.ensure_future(asyncio.sleep(ping_timeout))
 
-                # Return immediately if a ping timeout occurred.
-                if not timeout_handler.cancel() and timeout_handler.result():
+                done, pending = await asyncio.wait(
+                    [ping, wait], return_when=asyncio.FIRST_COMPLETED)
+
+                if ping in pending:
```

Comment:
Would probably be good to confirm the return value of ping is `PONG` just to be sure we actually got the right return from the server.